IllegalStateException: Preference node "org.eclipse.jdt.core" has been removed.

Recently I often get many errors like the one below saying `Preference node "org.eclipse.jdt.core" has been removed.`
The stack-trace is often different, coming from different code-paths but always ends at `org.eclipse.jdt.internal.core.JavaModelManager.getOption(JavaModelManager.java:2311)` trying to read from a removed preference node.

The affected project don't have project specific settings and therefore no .settings/*jdt.*.prefs file. But that worked well in the past.
The errors mainly show up while interactive rebasing but then can be annoying because a lot of things start to fail and I get multiple windows popping up warning me about failures for example in the clean-up actions or while reconceiling the AST.

The Eclipse-IDE is Version: 2022-06 M1 (4.24.0 M1), Build id: 20220414-1707

```
Java Model Exception: java.lang.IllegalStateException: Preference node "org.eclipse.jdt.core" has been removed.
	at org.eclipse.jdt.internal.core.CompilationUnitProblemFinder.process(CompilationUnitProblemFinder.java:326)
	at org.eclipse.jdt.internal.core.CompilationUnitProblemFinder.process(CompilationUnitProblemFinder.java:348)
	at org.eclipse.jdt.internal.core.ReconcileWorkingCopyOperation.makeConsistent(ReconcileWorkingCopyOperation.java:193)
	at org.eclipse.jdt.internal.core.ReconcileWorkingCopyOperation.executeOperation(ReconcileWorkingCopyOperation.java:94)
	at org.eclipse.jdt.internal.core.JavaModelOperation.run(JavaModelOperation.java:740)
	at org.eclipse.jdt.internal.core.JavaModelOperation.runOperation(JavaModelOperation.java:806)
	at org.eclipse.jdt.internal.core.CompilationUnit.reconcile(CompilationUnit.java:1319)
	at org.eclipse.jdt.internal.ui.text.java.JavaReconcilingStrategy.reconcile(JavaReconcilingStrategy.java:132)
	at org.eclipse.jdt.internal.ui.text.java.JavaReconcilingStrategy$1.run(JavaReconcilingStrategy.java:94)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:45)
	at org.eclipse.jdt.internal.ui.text.java.JavaReconcilingStrategy.reconcile(JavaReconcilingStrategy.java:91)
	at org.eclipse.jdt.internal.ui.text.java.JavaReconcilingStrategy.reconcile(JavaReconcilingStrategy.java:158)
	at org.eclipse.jdt.internal.ui.text.CompositeReconcilingStrategy.reconcile(CompositeReconcilingStrategy.java:94)
	at org.eclipse.jdt.internal.ui.text.JavaCompositeReconcilingStrategy.reconcile(JavaCompositeReconcilingStrategy.java:107)
	at org.eclipse.jface.text.reconciler.MonoReconciler.process(MonoReconciler.java:76)
	at org.eclipse.jface.text.reconciler.AbstractReconciler$BackgroundThread.run(AbstractReconciler.java:207)
Caused by: java.lang.IllegalStateException: Preference node "org.eclipse.jdt.core" has been removed.
	at org.eclipse.core.internal.preferences.EclipsePreferences.checkRemoved(EclipsePreferences.java:165)
	at org.eclipse.core.internal.resources.ProjectPreferences.internalGet(ProjectPreferences.java:495)
	at org.eclipse.core.internal.preferences.EclipsePreferences.get(EclipsePreferences.java:444)
	at org.eclipse.jdt.internal.core.JavaModelManager.getOption(JavaModelManager.java:2311)
	at org.eclipse.jdt.internal.core.JavaProject.getOption(JavaProject.java:2148)
	at org.eclipse.jdt.internal.core.AbstractClassFile.validateClassFile(AbstractClassFile.java:416)
	at org.eclipse.jdt.internal.core.AbstractClassFile.validateExistence(AbstractClassFile.java:498)
	at org.eclipse.jdt.internal.core.Openable.generateInfos(Openable.java:252)
	at org.eclipse.jdt.internal.core.JavaElement.openWhenClosed(JavaElement.java:597)
	at org.eclipse.jdt.internal.core.JavaElement.getElementInfo(JavaElement.java:328)
	at org.eclipse.jdt.internal.core.JavaElement.getElementInfo(JavaElement.java:314)
	at org.eclipse.jdt.internal.core.JavaElement.exists(JavaElement.java:231)
	at org.eclipse.jdt.internal.core.Openable.exists(Openable.java:208)
	at org.eclipse.jdt.internal.core.ModularClassFile.exists(ModularClassFile.java:317)
	at org.eclipse.jdt.internal.core.PackageFragmentRoot.getSourceModuleDescription(PackageFragmentRoot.java:927)
	at org.eclipse.jdt.internal.core.PackageFragmentRoot.getModuleDescription(PackageFragmentRoot.java:904)
	at org.eclipse.jdt.internal.core.JarPackageFragmentRoot.getModuleDescription(JarPackageFragmentRoot.java:307)
	at org.eclipse.jdt.internal.core.NameLookup.getModuleDescription(NameLookup.java:944)
	at org.eclipse.jdt.internal.core.SearchableEnvironment.getModuleDescription(SearchableEnvironment.java:1085)
	at org.eclipse.jdt.internal.core.SearchableEnvironment.getModulesDeclaringPackage(SearchableEnvironment.java:1015)
	at org.eclipse.jdt.internal.compiler.env.IModuleAwareNameEnvironment.getUniqueModulesDeclaringPackage(IModuleAwareNameEnvironment.java:113)
	at org.eclipse.jdt.internal.compiler.lookup.LookupEnvironment.computePackageFrom(LookupEnvironment.java:745)
	at org.eclipse.jdt.internal.compiler.lookup.LookupEnvironment.getTypeFromCompoundName(LookupEnvironment.java:1771)
	at org.eclipse.jdt.internal.compiler.lookup.LookupEnvironment.getTypeFromConstantPoolName(LookupEnvironment.java:1821)
	at org.eclipse.jdt.internal.compiler.lookup.LookupEnvironment.getTypeFromConstantPoolName(LookupEnvironment.java:1829)
	at org.eclipse.jdt.internal.compiler.lookup.LookupEnvironment.getTypeFromSignature(LookupEnvironment.java:1899)
	at org.eclipse.jdt.internal.compiler.lookup.BinaryTypeBinding.createFields(BinaryTypeBinding.java:860)
	at org.eclipse.jdt.internal.compiler.lookup.BinaryTypeBinding.cachePartsFrom(BinaryTypeBinding.java:610)
	at org.eclipse.jdt.internal.compiler.lookup.LookupEnvironment.createBinaryTypeFrom(LookupEnvironment.java:1055)
	at org.eclipse.jdt.internal.compiler.lookup.LookupEnvironment.createBinaryTypeFrom(LookupEnvironment.java:1036)
	at org.eclipse.jdt.internal.compiler.Compiler.accept(Compiler.java:308)
	at org.eclipse.jdt.internal.compiler.lookup.LookupEnvironment.askForType(LookupEnvironment.java:257)
	at org.eclipse.jdt.internal.compiler.lookup.UnresolvedReferenceBinding.resolve(UnresolvedReferenceBinding.java:114)
	at org.eclipse.jdt.internal.compiler.lookup.BinaryTypeBinding.resolveType(BinaryTypeBinding.java:248)
	at org.eclipse.jdt.internal.compiler.lookup.PackageBinding.getTypeOrPackage(PackageBinding.java:252)
	at org.eclipse.jdt.internal.compiler.lookup.CompilationUnitScope.findImport(CompilationUnitScope.java:551)
	at org.eclipse.jdt.internal.compiler.lookup.CompilationUnitScope.findSingleImport(CompilationUnitScope.java:623)
	at org.eclipse.jdt.internal.compiler.lookup.CompilationUnitScope.faultInImports(CompilationUnitScope.java:452)
	at org.eclipse.jdt.internal.compiler.lookup.Scope.getBinding(Scope.java:2221)
	at org.eclipse.jdt.internal.compiler.ast.FieldDeclaration.resolve(FieldDeclaration.java:235)
	at org.eclipse.jdt.internal.compiler.lookup.FieldBinding.constant(FieldBinding.java:223)
	at org.eclipse.jdt.internal.compiler.ast.FieldReference.generateCode(FieldReference.java:254)
	at org.eclipse.jdt.internal.compiler.ast.Expression.generateOptimizedStringConcatenation(Expression.java:893)
	at org.eclipse.jdt.internal.compiler.codegen.CodeStream.generateStringConcatenationAppend(CodeStream.java:2635)
	at org.eclipse.jdt.internal.compiler.ast.BinaryExpression.generateCode(BinaryExpression.java:157)
	at org.eclipse.jdt.internal.compiler.ast.ConditionalExpression.generateCode(ConditionalExpression.java:313)
	at org.eclipse.jdt.internal.compiler.ast.Statement.generateArguments(Statement.java:443)
	at org.eclipse.jdt.internal.compiler.ast.AllocationExpression.generateCode(AllocationExpression.java:212)
	at org.eclipse.jdt.internal.compiler.ast.ReturnStatement.generateCode(ReturnStatement.java:211)
	at org.eclipse.jdt.internal.compiler.ast.AbstractMethodDeclaration.generateCode(AbstractMethodDeclaration.java:354)
	at org.eclipse.jdt.internal.compiler.ast.AbstractMethodDeclaration.generateCode(AbstractMethodDeclaration.java:291)
	at org.eclipse.jdt.internal.compiler.ast.TypeDeclaration.generateCode(TypeDeclaration.java:761)
	at org.eclipse.jdt.internal.compiler.ast.TypeDeclaration.generateCode(TypeDeclaration.java:831)
	at org.eclipse.jdt.internal.compiler.ast.CompilationUnitDeclaration.generateCode(CompilationUnitDeclaration.java:412)
	at org.eclipse.jdt.internal.compiler.Compiler.resolve(Compiler.java:1072)
	at org.eclipse.jdt.internal.compiler.Compiler.resolve(Compiler.java:1108)
	at org.eclipse.jdt.internal.core.CompilationUnitProblemFinder.process(CompilationUnitProblemFinder.java:282)
	... 15 more
Caused by: java.lang.IllegalStateException: Preference node "org.eclipse.jdt.core" has been removed.
	at org.eclipse.core.internal.preferences.EclipsePreferences.checkRemoved(EclipsePreferences.java:165)
	at org.eclipse.core.internal.resources.ProjectPreferences.internalGet(ProjectPreferences.java:495)
	at org.eclipse.core.internal.preferences.EclipsePreferences.get(EclipsePreferences.java:444)
	at org.eclipse.jdt.internal.core.JavaModelManager.getOption(JavaModelManager.java:2311)
	at org.eclipse.jdt.internal.core.JavaProject.getOption(JavaProject.java:2148)
	at org.eclipse.jdt.internal.core.AbstractClassFile.validateClassFile(AbstractClassFile.java:416)
	at org.eclipse.jdt.internal.core.AbstractClassFile.validateExistence(AbstractClassFile.java:498)
	at org.eclipse.jdt.internal.core.Openable.generateInfos(Openable.java:252)
	at org.eclipse.jdt.internal.core.JavaElement.openWhenClosed(JavaElement.java:597)
	at org.eclipse.jdt.internal.core.JavaElement.getElementInfo(JavaElement.java:328)
	at org.eclipse.jdt.internal.core.JavaElement.getElementInfo(JavaElement.java:314)
	at org.eclipse.jdt.internal.core.JavaElement.exists(JavaElement.java:231)
	at org.eclipse.jdt.internal.core.Openable.exists(Openable.java:208)
	at org.eclipse.jdt.internal.core.ModularClassFile.exists(ModularClassFile.java:317)
	at org.eclipse.jdt.internal.core.PackageFragmentRoot.getSourceModuleDescription(PackageFragmentRoot.java:927)
	at org.eclipse.jdt.internal.core.PackageFragmentRoot.getModuleDescription(PackageFragmentRoot.java:904)
	at org.eclipse.jdt.internal.core.JarPackageFragmentRoot.getModuleDescription(JarPackageFragmentRoot.java:307)
	at org.eclipse.jdt.internal.core.NameLookup.getModuleDescription(NameLookup.java:944)
	at org.eclipse.jdt.internal.core.SearchableEnvironment.getModuleDescription(SearchableEnvironment.java:1085)
	at org.eclipse.jdt.internal.core.SearchableEnvironment.getModulesDeclaringPackage(SearchableEnvironment.java:1015)
	at org.eclipse.jdt.internal.compiler.env.IModuleAwareNameEnvironment.getUniqueModulesDeclaringPackage(IModuleAwareNameEnvironment.java:113)
	at org.eclipse.jdt.internal.compiler.lookup.LookupEnvironment.computePackageFrom(LookupEnvironment.java:745)
	at org.eclipse.jdt.internal.compiler.lookup.LookupEnvironment.getTypeFromCompoundName(LookupEnvironment.java:1771)
	at org.eclipse.jdt.internal.compiler.lookup.LookupEnvironment.getTypeFromConstantPoolName(LookupEnvironment.java:1821)
	at org.eclipse.jdt.internal.compiler.lookup.LookupEnvironment.getTypeFromConstantPoolName(LookupEnvironment.java:1829)
	at org.eclipse.jdt.internal.compiler.lookup.LookupEnvironment.getTypeFromSignature(LookupEnvironment.java:1899)
	at org.eclipse.jdt.internal.compiler.lookup.BinaryTypeBinding.createFields(BinaryTypeBinding.java:860)
	at org.eclipse.jdt.internal.compiler.lookup.BinaryTypeBinding.cachePartsFrom(BinaryTypeBinding.java:610)
	at org.eclipse.jdt.internal.compiler.lookup.LookupEnvironment.createBinaryTypeFrom(LookupEnvironment.java:1055)
	at org.eclipse.jdt.internal.compiler.lookup.LookupEnvironment.createBinaryTypeFrom(LookupEnvironment.java:1036)
	at org.eclipse.jdt.internal.compiler.Compiler.accept(Compiler.java:308)
	at org.eclipse.jdt.internal.compiler.lookup.LookupEnvironment.askForType(LookupEnvironment.java:257)
	at org.eclipse.jdt.internal.compiler.lookup.UnresolvedReferenceBinding.resolve(UnresolvedReferenceBinding.java:114)
	at org.eclipse.jdt.internal.compiler.lookup.BinaryTypeBinding.resolveType(BinaryTypeBinding.java:248)
	at org.eclipse.jdt.internal.compiler.lookup.PackageBinding.getTypeOrPackage(PackageBinding.java:252)
	at org.eclipse.jdt.internal.compiler.lookup.CompilationUnitScope.findImport(CompilationUnitScope.java:551)
	at org.eclipse.jdt.internal.compiler.lookup.CompilationUnitScope.findSingleImport(CompilationUnitScope.java:623)
	at org.eclipse.jdt.internal.compiler.lookup.CompilationUnitScope.faultInImports(CompilationUnitScope.java:452)
	at org.eclipse.jdt.internal.compiler.lookup.Scope.getBinding(Scope.java:2221)
	at org.eclipse.jdt.internal.compiler.ast.FieldDeclaration.resolve(FieldDeclaration.java:235)
	at org.eclipse.jdt.internal.compiler.lookup.FieldBinding.constant(FieldBinding.java:223)
	at org.eclipse.jdt.internal.compiler.ast.FieldReference.generateCode(FieldReference.java:254)
	at org.eclipse.jdt.internal.compiler.ast.Expression.generateOptimizedStringConcatenation(Expression.java:893)
	at org.eclipse.jdt.internal.compiler.codegen.CodeStream.generateStringConcatenationAppend(CodeStream.java:2635)
	at org.eclipse.jdt.internal.compiler.ast.BinaryExpression.generateCode(BinaryExpression.java:157)
	at org.eclipse.jdt.internal.compiler.ast.ConditionalExpression.generateCode(ConditionalExpression.java:313)
	at org.eclipse.jdt.internal.compiler.ast.Statement.generateArguments(Statement.java:443)
	at org.eclipse.jdt.internal.compiler.ast.AllocationExpression.generateCode(AllocationExpression.java:212)
	at org.eclipse.jdt.internal.compiler.ast.ReturnStatement.generateCode(ReturnStatement.java:211)
	at org.eclipse.jdt.internal.compiler.ast.AbstractMethodDeclaration.generateCode(AbstractMethodDeclaration.java:354)
	at org.eclipse.jdt.internal.compiler.ast.AbstractMethodDeclaration.generateCode(AbstractMethodDeclaration.java:291)
	at org.eclipse.jdt.internal.compiler.ast.TypeDeclaration.generateCode(TypeDeclaration.java:761)
	at org.eclipse.jdt.internal.compiler.ast.TypeDeclaration.generateCode(TypeDeclaration.java:831)
	at org.eclipse.jdt.internal.compiler.ast.CompilationUnitDeclaration.generateCode(CompilationUnitDeclaration.java:412)
	at org.eclipse.jdt.internal.compiler.Compiler.resolve(Compiler.java:1072)
	at org.eclipse.jdt.internal.compiler.Compiler.resolve(Compiler.java:1108)
	at org.eclipse.jdt.internal.core.CompilationUnitProblemFinder.process(CompilationUnitProblemFinder.java:282)
	at org.eclipse.jdt.internal.core.CompilationUnitProblemFinder.process(CompilationUnitProblemFinder.java:348)
	at org.eclipse.jdt.internal.core.ReconcileWorkingCopyOperation.makeConsistent(ReconcileWorkingCopyOperation.java:193)
	at org.eclipse.jdt.internal.core.ReconcileWorkingCopyOperation.executeOperation(ReconcileWorkingCopyOperation.java:94)
	at org.eclipse.jdt.internal.core.JavaModelOperation.run(JavaModelOperation.java:740)
	at org.eclipse.jdt.internal.core.JavaModelOperation.runOperation(JavaModelOperation.java:806)
	at org.eclipse.jdt.internal.core.CompilationUnit.reconcile(CompilationUnit.java:1319)
	at org.eclipse.jdt.internal.ui.text.java.JavaReconcilingStrategy.reconcile(JavaReconcilingStrategy.java:132)
	at org.eclipse.jdt.internal.ui.text.java.JavaReconcilingStrategy$1.run(JavaReconcilingStrategy.java:94)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:45)
	at org.eclipse.jdt.internal.ui.text.java.JavaReconcilingStrategy.reconcile(JavaReconcilingStrategy.java:91)
	at org.eclipse.jdt.internal.ui.text.java.JavaReconcilingStrategy.reconcile(JavaReconcilingStrategy.java:158)
	at org.eclipse.jdt.internal.ui.text.CompositeReconcilingStrategy.reconcile(CompositeReconcilingStrategy.java:94)
	at org.eclipse.jdt.internal.ui.text.JavaCompositeReconcilingStrategy.reconcile(JavaCompositeReconcilingStrategy.java:107)
	at org.eclipse.jface.text.reconciler.MonoReconciler.process(MonoReconciler.java:76)
	at org.eclipse.jface.text.reconciler.AbstractReconciler$BackgroundThread.run(AbstractReconciler.java:207)
```

--- a/org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/JavaProject.java
+++ b/org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/JavaProject.java
@@ -1918,7 +1918,10 @@ public class JavaProject
 		if (!JavaProject.hasJavaNature(this.project)) return null;
 		// Get cached preferences if exist
 		JavaModelManager.PerProjectInfo perProjectInfo = JavaModelManager.getJavaModelManager().getPerProjectInfo(this.project, true);
-		if (perProjectInfo.preferences != null) return perProjectInfo.preferences;
+		IEclipsePreferences preferences = perProjectInfo.preferences;
+		if (checkPreferencesExist(preferences)) {
+			return preferences;
+		}
 		// Init project preferences
 		IScopeContext context = new ProjectScope(getProject());
 		final IEclipsePreferences eclipsePreferences = context.getNode(JavaCore.PLUGIN_ID);
@@ -1998,6 +2001,23 @@ public class JavaProject
 		};
 		eclipsePreferences.addPreferenceChangeListener(this.preferencesChangeListener);
 		return eclipsePreferences;
+	}
+
+	private boolean checkPreferencesExist(IEclipsePreferences preferences) {
+		if (preferences == null) {
+			return false;
+		}
+		try {
+			// check the root node ("")
+			if (preferences.nodeExists("")) { //$NON-NLS-1$
+				return true;
+			}
+		} catch (BackingStoreException e1) {
+			// shouldn't happen, but if happens, we continue below
+		}
+		JavaModelManager manager = JavaModelManager.getJavaModelManager();
+		manager.resetProjectPreferences(this);
+		return false;
 	}
 
 	@Override
@@ -2145,7 +2165,17 @@ public class JavaProject
 	 */
 	@Override
 	public String getOption(String optionName, boolean inheritJavaCoreOptions) {
-		return JavaModelManager.getJavaModelManager().getOption(optionName, inheritJavaCoreOptions, getEclipsePreferences());
+		IEclipsePreferences preferences = getEclipsePreferences();
+		JavaModelManager manager = JavaModelManager.getJavaModelManager();
+		String option;
+		try {
+			option = manager.getOption(optionName, inheritJavaCoreOptions, preferences);
+		} catch (IllegalStateException e) {
+			// preferences deleted right after the check? let's retry once
+			preferences = getEclipsePreferences();
+			option = manager.getOption(optionName, inheritJavaCoreOptions, preferences);
+		}
+		return option;
 	}
 
 	/**

